management-api-agent-common/src/main/java/io/k8ssandra/metrics/builder/CassandraMetricRegistryListener.java

```diff
@@ -0,0 +1,300 @@
+package io.k8ssandra.metrics.builder;
+
+import com.codahale.metrics.*;
+import io.k8ssandra.metrics.builder.filter.CassandraMetricDefinitionFilter;
+import io.prometheus.client.Collector;
+import org.apache.cassandra.utils.EstimatedHistogram;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import static io.k8ssandra.metrics.builder.CassandraMetricsTools.PRECOMPUTED_QUANTILES;
+import static io.k8ssandra.metrics.builder.CassandraMetricsTools.PRECOMPUTED_QUANTILES_TEXT;
+
+public class CassandraMetricRegistryListener implements MetricRegistryListener {
+
+    private static final org.slf4j.Logger logger = LoggerFactory.getLogger(CassandraMetricRegistryListener.class);
+
+    private final CassandraMetricNameParser parser;
+
+    private final CassandraMetricDefinitionFilter metricFilter;
+
+    private final ConcurrentHashMap<String, RefreshableMetricFamilySamples> familyCache;
+
+    // This cache is used for the remove purpose, we need dropwizardName -> metricName mapping
+    private final ConcurrentHashMap<String, String> cache;
+
+    public CassandraMetricRegistryListener(ConcurrentHashMap<String, RefreshableMetricFamilySamples> familyCache, CassandraMetricDefinitionFilter metricFilter) {
+        parser = new CassandraMetricNameParser(CassandraMetricsTools.DEFAULT_LABEL_NAMES, CassandraMetricsTools.DEFAULT_LABEL_VALUES);
+        cache = new ConcurrentHashMap<>();
+        this.familyCache = familyCache;
+        this.metricFilter = metricFilter;
+    }
+
+    public void updateCache(String dropwizardName, String metricName, RefreshableMetricFamilySamples prototype) {
+        prototype.getDefinitions().removeIf(next -> !metricFilter.matches(next, dropwizardName));
+
+        if (prototype.getDefinitions().size() < 1) {
+            return;
+        }
+
+        RefreshableMetricFamilySamples familySamples;
+        if (!familyCache.containsKey(metricName)) {
+            familyCache.put(metricName, prototype);
+            cache.put(dropwizardName, metricName);
```

Comment:
You could do it multiple times, sure. But I don't see how it would conflict. Why would same metric be registered twice in different threads? And if they're different, which one would be correct? If they're the same, then no problem.